### ISSUE TITLE
Support for AWS ECR

### ISSUE THREAD

**Issue by greysteil** [dependabot/dependabot-core#405]
_From @edkellena on February 1, 2018 9:21_

Dependabot is complaining that it can't authenticate to our ECR docker registry and we should add the login details to the dashboard.  However AWS works differently and you need to run their CLI to generate your login creds (which are time-limited).  
Something like `${aws ecr get-login --no-include-email --region $REGION}` 

An interesting problem for you to solve.  We could add IAM credentials into the dashboard to allow you to do this??

_Copied from original issue: dependabot/feedback#81_

**Comment by fernandocarletti** [comment]
Hello. Is there any love for this issue?

**Comment by greysteil** [comment]
There is and in fact @nesl247 tweeted about it a few hours ago.

I'm keen on this but need to do some research on how it would work.

**Comment by nesl247** [comment]
@fernandocarletti and I had just been discussing it prior to this (both work for the same company).

I wonder if a possible quick solution would be to allow the registry configuration in dependabot to be accessible via an API. We could then run a cronjob that calls that API with updated credentials every 11 hours or whatever.

That is until a more permanent solution could come about. Is there an API for updating it?

**Comment by nesl247** [comment]
@greysteil I've configured our dependabot with a username and password that should work with AWS, however, it still doesn't appear to be finding any updates.

How does dependabot check to see what newer tags there are?

**Comment by greysteil** [comment]
@nesl247 I'm crunching on getting Terraform support out this evening, but am planning to focus on Docker tomorrow. Barring a sea of bugs I'll get this sorted then 🙂

**Comment by nesl247** [comment]
So I'm thinking, and hoping I'm wrong, but it may be that AWS does not fully support the registry API. This is hard for me to debug since I don't know how to. I have written a small script that I can execute that will actually connect to api.dependabot.com and provide a docker registry entry with a username and password that should work for pulling.

However, dependabot is not finding any new tags for one of our applications. In my case, the tags are short commit references like `62c0ade`.

Sounds good @greysteil

**Comment by greysteil** [comment]
Ah, it may be that Dependabot just isn't picking those up because they're not version-like, and it wouldn't have a way to order it.

Have you had a look at [dependabot-script](https://github.com/dependabot/dependabot-script), to step through the logic there? It uses dependabot-core in basically the same way that the main Dependabot updater does, and is what I use to debug these issues if/when I need to step through the logic.

**Comment by nesl247** [comment]
I wasn't aware of dependabot-script, good to know.

I'm not sure how we should do our tags then. I was so focused on getting dependabot connected to AWS I wasn't thinking about that. In my head, it was like "well, git knows this is newer than the last..".

**Comment by nesl247** [comment]
We're switching to tagged releases. We were moving very fast, and skipped thinking about tagging for dependency docker images for us. So we'll keep this ticket back on track which is the support for ECR.

**Comment by greysteil** [comment]
I was hoping to get to this yesterday but spent the day on a couple of Python issues. It’s top of my list for the next few days.

**Comment by greysteil** [comment]
This is done! Sorry it took a while. That changes are in https://github.com/dependabot/dependabot-core/commit/e92672c2cb10a59e7646583bff60827222ccf099, and to use it you'll need to give Dependabot API keys for you AWS account (scoped using IAM to have read access only to ECR, of course).

I'm going to update the front-end to make it clearer how to add credentials.

Thanks for the feature request!

**Comment by nesl247** [comment]
Hey that's awesome! I guess we'll wait on the frontend as I am not quite sure how to do it.

**Comment by greysteil** [comment]
Frontend is added now (well, it's just some additional comments in the description of the fields under "config variables"). Happy to talk you through if not clear.

**Comment by nesl247** [comment]
I saw that after I commented and had hoped I got the comment changed in time, guess not 😄 .

**Comment by greysteil** [comment]
Looking into it!

**Comment by nesl247** [comment]
Unfortunately we have a new issue.

`build(deps): bump docker-php from 7.2-0.3 to 7.2-9658143 in /docker/cart-sync`

This should have used `7.2-0.3.1`. Instead, it's changed to the commit hashes.

**Comment by nesl247** [comment]
Thanks for the ECR support, that part is great. I'm not sure if my above comment should be made into a new issue or not. I'm doing a new response since I know how quickly you respond sometimes 😄

**Comment by greysteil** [comment]
Huh, that one is tricky! Did we just get unlucky here and have a commit sha version that happened to start with 7 digits (no letters)? If so I think the solution might be asking dependabot to ignore that specific version - I’ll make sure that’s possible for Docker now.

**Comment by nesl247** [comment]
It's quite possible. Most of our hashes are 7 digits though, so unless this is permanently fixed, we're back to not using dependabot, or we stop using tags, which we started because of our previous discussion.

**Comment by greysteil** [comment]
That's fine - I can put something in that stops it from switching from more traditional looking versions to 7 digit numbers. I'll do that now.

**Comment by nesl247** [comment]
Sorry, I think I sounded a bit snippy, not my intention.

I think as long as the current tag is a version, the logic should be to not switch to commit hashes, even if it is newer. I would think the only time to switch to commit hashes if it it is currently using latest, or it already uses a hash.

**Comment by greysteil** [comment]
Oh, hang on, I might be able to do one better. Give me a moment...

**Comment by greysteil** [comment]
Yep, have made it work without the need to change the prefix to be non-numeric. Deploying now.

**Comment by nesl247** [comment]
It's working beautifully. Thank you so much for the fast turnaround.

**Comment by greysteil** [comment]
No worries - all helps to make Dependabot better! 🎉

**Comment by nesl247** [comment]
One quick question, when using the api for updating private deps, do we have to specify `docker-php` for example as the dependency name, or the full namespace of the docker image? For example, `something.something.aws.amazon.com/docker-php`?

**Comment by greysteil** [comment]
Just `docker-php` (or whatever the name in the PR title you're seeing is `bump <name> from ...`). 🙂

**Comment by nesl247** [comment]
Awesome, thanks.
